hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationPublisherDelegate.java

```diff
@@ -16,53 +16,160 @@
 
 package com.hazelcast.wan.impl;
 
+import com.hazelcast.internal.services.ServiceNamespace;
+import com.hazelcast.monitor.LocalWanPublisherStats;
+import com.hazelcast.spi.partition.PartitionReplicationEvent;
 import com.hazelcast.wan.WanReplicationEvent;
-import com.hazelcast.wan.WanReplicationEndpoint;
 import com.hazelcast.wan.WanReplicationPublisher;
 
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.util.MapUtil.createHashMap;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
 /**
- * Delegating replication publisher implementation
+ * Delegating WAN replication publisher implementation. This implementation
+ * is a container for multiple WAN publishers.
+ * When publishing an event on this delegate, all publishers are notified.
  */
-final class WanReplicationPublisherDelegate
-        implements WanReplicationPublisher {
-
+public final class WanReplicationPublisherDelegate {
+    /** Non-null WAN replication name */
     final String name;
-    final WanReplicationEndpoint[] endpoints;
+    /** Non-null WAN publishers, grouped by publisher ID */
+    final ConcurrentMap<String, WanReplicationPublisher> publishers;
 
-    WanReplicationPublisherDelegate(String name, WanReplicationEndpoint[] endpoints) {
+    public WanReplicationPublisherDelegate(String name,
+                                           ConcurrentMap<String, WanReplicationPublisher> publishers) {
+        checkNotNull(name, "WAN publisher name should not be null");
+        checkNotNull(publishers, "WAN publisher map should not be null");
         this.name = name;
-        this.endpoints = endpoints;
+        this.publishers = publishers;
     }
 
-    public WanReplicationEndpoint[] getEndpoints() {
-        return endpoints;
+    /** Returns all {@link WanReplicationPublisher}s for this delegate */
+    public @Nonnull
+    Collection<WanReplicationPublisher> getPublishers() {
+        return publishers.values();
+    }
+
+    /**
+     * Returns the {@link WanReplicationPublisher} with the {@code publisherId}
+     * or {@code null} if it doesn't exist.
+     */
+    public WanReplicationPublisher getPublisher(String publisherId) {
+        return publishers.get(publisherId);
+    }
+
+    public void addPublisher(String publisherId,
```

Comment:
is lack of idempotency an issue here? Is it ok to add different publisher object with same id?

Reply:
We shouldn't have publishers with the same publisherId. I'll check if this is the case and add some checks.